Improve base padding rule

benkimpel opened the pull request:
- Move AST utility functions to separate file
- Create/export enums to help build/read configuration objects
- Add an interface and better property names for configuration objects
- Rename `makeRule` to `createRule` and move to main file
- Delete `src/utils.ts`
- Remove isTestFile and use `overrides.files` to specify lint targets
- Fix last implicit any and enable `noImplicitAny` in tsconfig
- Add a padding-around-all rule built from the others' configs
- Add missing x*/f* Jest functions

Closes #50
Closes #51

benkimpel commented:
Ok, _now_ it's ready for a look, @dangreenisrael.

Sorry for the large commit. Given there was a refactor alongside additions, it was too hard to break this into several smaller commits.

I'd recommend skipping the diff on `src/rules/padding.ts` and just reading through the file as a whole. I added a huge comment at the bottom to explain the algorithm, moved a lot of functions to module scope, renamed functions and parameterss, added comments and types for clarity, etc.

benkimpel commented on src/rules/padding.ts:
```diff
@@ -1,115 +1,142 @@
- * @fileoverview Rule to require or disallow newlines between jest functions
- *   Based on eslint/padding-line-between-statements by Toru Nagashima
- *   See: https://github.com/eslint/eslint/blob/master/lib/rules/padding-line-between-statements.js
+ * Require/fix newlines between jest functions
```
s/between jest functions/around jest functions/

dangreenisrael commented on tsconfig.json:
```diff
@@ -1,7 +1,7 @@
   "compilerOptions": {
     "module": "commonjs",
-    "noImplicitAny": false,
+    "noImplicitAny": true,
```
nice!

benkimpel replied:
yeah, i'm excited about that one!!

dangreenisrael commented on src/ast-utils.ts:
```diff
@@ -0,0 +1,78 @@
+export const isSemicolonToken = (token: AST.Token): boolean =>
+  token.value === ';' && token.type === 'Punctuator';
+
+export const isTokenOnSameLine = (
```
Any interest in writing some unit tests for these utils. It isn't necessary as we have integration tests for all the rules - just a nice to have.

dangreenisrael commented on src/ast-utils.ts:
```diff
@@ -0,0 +1,78 @@
+export const isSemicolonToken = (token: AST.Token): boolean =>
+  token.value === ';' && token.type === 'Punctuator';
+
+export const isTokenOnSameLine = (
```
```suggestion
export const areTokensOnSameLine= (
```

benkimpel replied:
:+1:

dangreenisrael commented on src/ast-utils.ts:
```diff
@@ -0,0 +1,78 @@
+import { AST, SourceCode } from 'eslint';
+import { Node } from 'estree';
+
+export const isSemicolonToken = (token: AST.Token): boolean =>
```
```suggestion
export const isTokenASemicolon = (token: AST.Token): boolean =>
```

benkimpel replied:
I like it. I'm on it.

dangreenisrael commented on src/index.ts:
```diff
@@ -3,106 +3,260 @@
+    },
+    {
+      paddingType: PaddingType.Always,
+      prevStatementType: [
```
nice!

dangreenisrael commented on src/index.ts:
```diff
@@ -3,106 +3,260 @@
-      'jest-formatting/padding-around-before-each-blocks': 2,
-      'jest-formatting/padding-around-describe-blocks': 2,
-      'jest-formatting/padding-around-test-blocks': 2,
+    overrides: {
```
nice!

dangreenisrael commented on src/rules/padding.ts:
```diff
@@ -137,203 +164,180 @@ function verifyForAlways(
-      if (!validParent) {
-        return;
-      }
+  // This is basically... "eh, whatever"
```
This comment isn't super clear to me

benkimpel replied:
Oops. Yes... clear as mud. I'll fix!

dangreenisrael commented on src/rules/padding.ts:
```diff
@@ -137,203 +164,180 @@ function verifyForAlways(
+  //  ESTree.Node doesn't support the parent property which is added by
+  //  ESLint during traversal. Our best bet is to ignore the property access
+  //  here as it's the only place that it's checked.
+  // @ts-ignore
```
+1 to `@ts-ignore` instead of `any`

benkimpel commented:
@dangreenisrael I think I've addressed your feedback, but let me know if I've missed anything. I did _some_ work on test coverage for ast utils (ast-utils.spec.ts). Two functions left to cover, but i'd vote to handle them at a later time.

dangreenisrael commented:
@benkimpel Feel free to merge this

benkimpel commented:
@dangreenisrael merge coming up! gonna squish the PR feedback commit... not really valuable for historical purposes.
